src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -1605,12 +1609,75 @@ protected override void TransformManagedToNative(ILCodeStream codeStream)
 
 #if READYTORUN
             var stringToAnsi =
-                Context.SystemModule.GetKnownType("System.StubHelpers", "AnsiBSTRMarshaler")
+                Context.SystemModule.GetKnownType("System.StubHelpers", "CSTRMarshaler")
                 .GetKnownMethod("ConvertToNative", null);
+
+            bool bPassByValueInOnly = In && !Out && !IsManagedByRef;
+            
+            if (bPassByValueInOnly)
+            {
+                var bufSize = emitter.NewLocal(Context.GetWellKnownType(WellKnownType.Int32));
+                _localBuffer = emitter.NewLocal(Context.GetWellKnownType(WellKnownType.IntPtr));
+
+                // LocalBuffer = 0
+                codeStream.Emit(ILOpcode.ldnull);
+                codeStream.EmitStLoc(_localBuffer);
+
+                var noOptimize = emitter.NewCodeLabel();
+
+                // if == NULL, goto NoOptimize
+                LoadManagedValue(codeStream);
+                codeStream.Emit(ILOpcode.brfalse, noOptimize);
+
+                // String.Length + 2
+                LoadManagedValue(codeStream);
+                var stringLen =
+                            Context.SystemModule.GetKnownType("System", "String")
+                            .GetKnownMethod("get_Length", null);
+                codeStream.Emit(ILOpcode.call, emitter.NewToken(stringLen));
+                codeStream.EmitLdc(2);
+                codeStream.Emit(ILOpcode.add);
+
+                // (String.Length + 2) * GetMaxDBCSCharByteSize()
+                codeStream.Emit(ILOpcode.ldsfld, emitter.NewToken(Context.SystemModule.GetKnownType(
+                                                "System.Runtime.InteropServices","Marshal")
+                                                .GetKnownField("SystemMaxDBCSCharSize")));
+                codeStream.Emit(ILOpcode.mul_ovf);
+
+                // BufSize = (String.Length + 2) * GetMaxDBCSCharByteSize()
+                codeStream.EmitStLoc(bufSize);
+
+                // if (MAX_LOCAL_BUFFER_LENGTH < BufSize ) goto NoOptimize
+                codeStream.EmitLdc(MAX_LOCAL_BUFFER_LENGTH + 1); 
+                codeStream.EmitLdLoc(bufSize);
+                codeStream.Emit(ILOpcode.clt);
+                codeStream.Emit(ILOpcode.brtrue, noOptimize);
+
+                // LocalBuffer = localloc(BufSize);
+                codeStream.EmitLdLoc(bufSize);
+                codeStream.Emit(ILOpcode.localloc);
+                codeStream.EmitStLoc(_localBuffer);
+
+                // NoOptimize:
+                codeStream.EmitLabel(noOptimize);
+            }
+
             int flags = (PInvokeFlags.BestFitMapping ? 0x1 : 0)
                 | (PInvokeFlags.ThrowOnUnmappableChar ? 0x100 : 0);
+
+            // CSTRMarshaler.ConvertToNative pManaged, dwAnsiMarshalFlags, pLocalBuffer
             codeStream.EmitLdc(flags);
             LoadManagedValue(codeStream);
+
+            if (_localBuffer != default)
```

Comment:
Might be better to use `ILLocalVariable?` as the type of _localBuffer. This condition will also be true if we created a local and it was the first local (index 0).

Reply:
makes sense, changed in latest commit.